api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java

```diff
@@ -490,4 +514,93 @@ public FirecloudNihStatus getNihStatus() {
           }
         });
   }
+
+  @Override
+  public void addProjectToServicePerimeter(String servicePerimeterName, String billingProject) {
+    final String utf8 = StandardCharsets.UTF_8.name();
+
+    // yes this actually gets URL decoded twice
+    // TODO: update Rawls, then FC Orch, then AoU.  See Terra JIRA AS-559
+    final String doublyEncodedName;
+    try {
+      doublyEncodedName = URLEncoder.encode(URLEncoder.encode(servicePerimeterName, utf8), utf8);
+    } catch (UnsupportedEncodingException e) {
+      throw new ServerErrorException(e);
+    }
+
+    ServicePerimetersApi perimetersApi = servicePerimetersApiProvider.get();
+    retryHandler.run(
+        (context) -> {
+          try {
+            perimetersApi.addProjectToServicePerimeter(doublyEncodedName, billingProject);
+          } catch (ApiException e) {
+            String alreadyInMsg =
+                String.format(
+                    "project %s is already in service perimeter %s",
+                    billingProject, servicePerimeterName);
+            if (configProvider.get().featureFlags.badRequestIsOkWhenAlreadyInPerimeter
+                && e.getCode() == 400
+                && e.getResponseBody().contains(alreadyInMsg)) {
+              log.info(alreadyInMsg);
+            } else {
+              throw e;
+            }
+          }
+          return null;
+        });
+
+    // it takes some time to add the project to the perimeter
+    waitForReadyProject(billingProject);
+  }
+
+  private void waitForReadyProject(String billingProject) throws WorkbenchException {
```

Comment:
NOTE: This can take more than 15 seconds, which the user will experience as a create/duplicate workspace delay.

Reply:
I see - I didn't understand that this was an async state transition, I thought it was just a slow, but transactional API call. That complicates things.. I'm not sure I'm comfortable depending on a polling loop within this method - sorry, you may have been saying this all along but I didn't make the connection on this until now.

Do you have any other ideas for alternatives, apart from the double-buffer solution?

Reply:
Double Buffer might not be too bad, if we add a perimeter name (or Tier ID) column to the buffer table and filter on it for all buffer operations

An alternative to this polling loop might be to tolerate un-READY workspaces if they are ADDINGTOPERIMETER.  Not sure how feasible that is, but I can play with it a bit.

Reply:
Terra interprets Workspace Clone requests to unREADY projects as Project Not Found - I haven't tried Create but I assume it's the same backend logic.  So we'd need the same waitForReady polling, only slightly later in the process.

We could allow user interaction in AoU with a workspace before we create it in Terra.  But that seems very likely to explode in complexity and rework.  Let's not do that.